File: apps/desktop/src/utils/member.utils.ts
import { Member, Nullable } from "@repo/types";
import {
  getMemberExceedsLimits,
  getRec,
  TRIAL_DURATION_DAYS,
} from "@repo/utilities";
import { getIntl } from "../i18n";
import type { AppState } from "../state/app.state";
import { EffectivePlan } from "../types/member.types";

export const getMyMember = (state: AppState): Nullable<Member> => {
  return getRec(state.memberById, state.auth?.uid) ?? null;
};

export const getEffectivePlan = (state: AppState): EffectivePlan => {
  if (state.isEnterprise) {
    return "enterprise";
  }
  return getMyMember(state)?.plan ?? "community";
};

export const planToDisplayName = (plan: EffectivePlan): string => {
  if (plan === "enterprise") {
    return getIntl().formatMessage({ defaultMessage: "Enterprise" });
  } else if (plan === "community") {
    return getIntl().formatMessage({ defaultMessage: "Community" });
  } else if (plan === "free") {
    return getIntl().formatMessage({ defaultMessage: "Free" });
  } else {
    return getIntl().formatMessage({ defaultMessage: "Pro" });
  }
};

export const getIsOnTrial = (state: AppState): boolean => {
  const member = getMyMember(state);
  return member?.isOnTrial === true;
};

const TRIAL_DURATION_MS = TRIAL_DURATION_DAYS * 24 * 60 * 60 * 1000;

export const getTrialDaysRemaining = (state: AppState): number | null => {
  const member = getMyMember(state);
  if (!member?.isOnTrial || !member.trialEndsAt) {
    return null;
  }

  const now = Date.now();
  const endsAt = new Date(member.trialEndsAt).getTime();
  const msRemaining = endsAt - now;
  return Math.max(0, Math.ceil(msRemaining / (1000 * 60 * 60 * 24)));
};

export const getTrialProgress = (state: AppState): number | null => {
  const member = getMyMember(state);
  if (!member?.isOnTrial || !member.trialEndsAt) {
    return null;
  }

  const now = Date.now();
  const endsAt = new Date(member.trialEndsAt).getTime();
  const msRemaining = endsAt - now;
  return Math.max(0, Math.min(1, msRemaining / TRIAL_DURATION_MS));
};

export const getIsPro = (state: AppState): boolean => {
  const member = getMyMember(state);
  if (!member) {
    return false;
  }

  return member.plan !== "free";
};

export const getIsPaidSubscriber = (state: AppState): boolean => {
  return getIsPro(state) && !getIsOnTrial(state);
};

export const getMemberExceedsLimitByState = (state: AppState): boolean => {
  const member = getMyMember(state);
  const config = state.config;
  if (!member || !config) {
    return false;
  }

  return getMemberExceedsLimits(member, config);
};

File: mobile/lib/utils/member_utils.dart
import 'package:app/model/member_model.dart';
import 'package:app/state/app_state.dart';
import 'package:intl/intl.dart';

enum EffectivePlan { free, pro }

EffectivePlan getEffectivePlan(AppState state) {
  final member = state.member;
  if (member == null) return EffectivePlan.free;
  if (member.plan == .pro) return EffectivePlan.pro;
  return EffectivePlan.free;
}

const trialDurationDays = 7;
final _trialDurationMs = trialDurationDays * Duration.millisecondsPerDay;

bool getIsOnTrial(AppState state) {
  return state.member?.isOnTrial == true;
}

bool getIsPro(AppState state) {
  final member = state.member;
  if (member == null) return false;
  return member.plan != MemberPlan.free;
}

int? getTrialDaysRemaining(AppState state) {
  final member = state.member;
  if (member?.isOnTrial != true || member?.trialEndsAt == null) {
    return null;
  }

  final now = DateTime.now().millisecondsSinceEpoch;
  final endsAt = DateTime.parse(member!.trialEndsAt!).millisecondsSinceEpoch;
  final msRemaining = endsAt - now;
  final days = (msRemaining / Duration.millisecondsPerDay).ceil();
  return days < 0 ? 0 : days;
}

double? getTrialProgress(AppState state) {
  final member = state.member;
  if (member?.isOnTrial != true || member?.trialEndsAt == null) {
    return null;
  }

  final now = DateTime.now().millisecondsSinceEpoch;
  final endsAt = DateTime.parse(member!.trialEndsAt!).millisecondsSinceEpoch;
  final msRemaining = endsAt - now;
  final progress = msRemaining / _trialDurationMs;
  return progress.clamp(0.0, 1.0);
}

final _numberFormat = NumberFormat('#,###');

int? getFreeWordsRemaining(AppState state) {
  final config = state.config;
  final member = state.member;
  if (config == null || member == null) return null;
  final remaining = config.freeWordsPerDay - member.wordsToday;
  return remaining < 0 ? 0 : remaining;
}

double? getFreeWordsProgress(AppState state) {
  final config = state.config;
  final member = state.member;
  if (config == null || member == null || config.freeWordsPerDay == 0) {
    return null;
  }
  final used = member.wordsToday / config.freeWordsPerDay;
  return used.clamp(0.0, 1.0);
}

String formatWordCount(int count) => _numberFormat.format(count);
